### DISCUSSION TITLE
Atomic grouping

### DISCUSSION THREAD

**Discussion by intrnl** [windicss/windicss#253] [Ideas]
Had this idea that seems too good to pass up, and thought it might be worth to discuss it further by sharing it here.

## Introducing atomic grouping

The one thing that stands out the most when creating components using atomic CSS solutions like Tailwind is that it's very easy to end up with very long class names in markup.

Given that nowadays we mostly work with UI libraries like Svelte or Vue, with their respective single file components, I thought it might be good to solve the above issue by being able to group atomic classes, by putting them inside the `<style>` element, here's an example:

```svelte
<div class='nav'>
  <a href='#' class='nav-item'>Home</a>
  <a href='#' class='nav-item'>Posts</a>
  <a href='#' class='nav-item'>About</a>
</div>

<style atomic>
  .nav {
    @apply flex;
    @apply space-x-10;
    @apply px-5 py-4;
    @apply border-b-2 border-gray-100;
  }

  .nav-item {
    @apply text-gray-500;
    @apply hover:text-gray-900;
  }
</style>
```

The `@apply` directive here doesn't [extract styling](https://tailwindcss.com/docs/extracting-components), instead it merely acts as a redirection, resulting in this output:

```svelte
<div class='flex border-b-2 border-gray-100 px-5 py-4 space-x-10'>
  <a href='#' class='text-gray-500 hover:text-gray-900'>Home</a>
  <a href='#' class='text-gray-500 hover:text-gray-900'>Posts</a>
  <a href='#' class='text-gray-500 hover:text-gray-900'>About</a>
</div>
```

This makes atomic CSS a lot easier to reason with.

## Using multiple grouping

It's possible to use multiple grouping, for example:

```svelte
<div class='padded header'></div>
<div class='padded footer'></div>

<style atomic>
  .padded {
    @apply py-4 px-8;
  }

  .header {
    @apply bg-gray-100;
  }

  .footer {
    @apply bg-gray-300;
  }
</style>
```

## One-off styling

Another benefit I can see with this is that you can easily defer to regular CSS for one-off styling if you really need to.

```svelte
<div class='header'></div>

<style atomic>
  .header {
    @apply flex;
    background-color: hotpink;
  }
</style>
```

which would result in the following output:

```svelte
<div class='flex c-header'></div>

<style>
  .c-header {
    background-color: hotpink;
  }
</style>
```

## Known limitations

As `@apply` here only works as a redirection, it wouldn't work with anything else besides simple class selectors, so this won't work:

```svelte
<style atomic>
  .nav > .nav-item {
    /* This won't work! */
    @apply text-gray-500;
  }

  @media print {
    .nav-item {
      /* Won't work inside media queries either! */
      @apply text-black;
    }
  }
<style>
```

However, this is still pretty useful, being able to use regular CSS classes as a way of grouping atomic ones, and you can still do one-off styling for the above cases if you really need to.

## But why?

I think the most question asked when looking at this idea would be, "why? aren't you just reimplementing regular CSS?", to which I'd say that I don't think that's the case.

Atomic CSS has made it easier to reason with styling, no more having to care about specificity, redundancy and bloat, and this idea is just one way of improving the maintainability aspect of it.

**Comment by ahku** [comment]
We suggested something similar in this issue: https://github.com/windicss/windicss/issues/225

But I like this idea with atomic style! Very elegant. 

As for the limitation, perhaps classes in media queries can just become normal CSS with the selector:

```html
<style atomic>
.head {
  @apply py-4 sm:py-6;

  @media print {
    @apply text-red-500;
  }
}
</style>

<!-- output -->
<template>
<div class="header py-4 sm:py-6"></div>
<template>

<style>
@media print {
  .header { color: red}
}
</style>
```

Above the header selector remain due to the css, but maybe it's good to include it regardless?

I'm thinking though, wouldn't it be possible to apply the variants if we use the variant queries:


```
@screen sm {
  @apply py-6; 
}
```

Should become:

```
<div class="sm:py-6"></div>
```

**Reply by ahku** [reply]
@intrnl Thought some more on this, and although I live the `<style atomic>` notion, I think it would be too much tied to Vue. 

Wouldn't it be better to create an @atomic or @extend (coming from SCSS and LESS):

```css
.header {
  @extend py-4 sm:py-6;
}
```

That way, this can be also be applied inside CSS files.

The drawbacks will be the same as with @apply command, or any CSS: There might be multiple `.header` selectors all over the code base. But now you can also use it in Vue SFCs in combination with `<style scoped>`. See behaviour in my previous reply.

The biggest problem here though, is maybe mutating the HTML… probably possible with Vite-plugins and Vue/React, but the technique won't be useful when styling static HTML. But then we could fallback to @apply behaviour…

@voorjaar @antfu What do you think about this? I think could be a huge improvement. It would allow people to either write utility classes directly in HTML, or write selectors and still get the advantages of utility classes.

**Reply by intrnl** [reply]
I'm not too sure about having `@apply` return to the previous behavior of extracting styling when used on an unsupported selector, I think it might cause confusion over what's redirection and what's extraction. It might be better to throw as is, now that the `<style>` element is in a special mode.

As for using grouping in places outside Svelte/Vue SFC, I'm not too sure on that one either, but if it's just wanting to link to an external file, then it's easily doable through the `src` attribute.

```svelte
<style atomic src='../styles/base.css' />
```

**Reply by ahku** [reply]
It might be more confusing that `@apply` suddenly gets a new behaviour, even if the reason is that "it's wrapped inside an atomic style tag". And it also limits the "redirection" to be used inside style tags, ie. you can't use it in pure css-files.

But if we create a new command, `@extend`, which has a historic meaning of grouping classes in SCSS/LESS, then we can implement the redirection/atomic operation both in style tags and css-files (without altering how we include them into our codebase)

Here are some examples, which combines `@extend`, `@apply`, and normal CSS:

## CSS file
``` html
<!-- HTML -->
<div class="header"></div>
<div class="footer"></div>
```

```css
/* style.css */
.header {
  @extend py-4 sm:py-6;
  @apply bg-gray-200;
  background-image: url('image');
}
.footer {
  @extend text-xs sm:text-sm;
}
```

### Output:
``` html
<!-- HTML -->
<div class="header py-4 sm:py-6"></div>
<div class="text-sm sm:text-sm"></div>
```

```css
/* style.css */
.header {
  background-color: rgba(229, 229, 229, var(--tw-bg-opacity));
  background-image: url('image');
}
```


## Vue SFC, with scoped styles
```html
<template>
<div class="header"></div>
</template>

<style scoped>
.header {
  @extend py-4 sm:py-6;
  @apply bg-gray-200;
  background-image: url('image');
}
</style>
````

### Output:
```html
<div class="header py-4 sm:py-6" data-v-2f5f086e></div>

<style>
.header[data-v-2f5f086e] {
  background-color: rgba(229, 229, 229, var(--tw-bg-opacity));
  background-image: url('image');
}
</style>
```

**Reply by ahku** [reply]
@intrnl I guess this idea is redundant now with the upcoming [Alias config](https://windicss.org/posts/v30.html#alias-config) feature! 

And in combination with the future [Windi language function](https://windicss.org/posts/language.html#alias), our wishes have come true!

So excited for these new additions!
